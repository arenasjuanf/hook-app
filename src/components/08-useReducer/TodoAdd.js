import React from 'react';
import {useForm } from '../../hooks/useForm'


export const TodoAdd = ({ handleAddTodo }) => {

        
    const [{description}, handleInputChanges, reset ] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(!description || !description.trim()){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();
    }


    return (
        <>
            <h4 className="text-center add-title">
                Add TODO
            </h4>
            <form onSubmit={handleSubmit}>
                <input
                    className="form-control w-100"
                    type="text"
                    name="description"
                    placeholder="Example: Do homeworks"
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChanges}
                />

                <button type="submit" className="btn btn-outline-primary mt-2 btn-block w-100">
                    Add
                </button>
            </form>  
        </>
    )
}
