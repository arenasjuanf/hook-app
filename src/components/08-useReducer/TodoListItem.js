import React from 'react';
import propTypes from 'prop-types';

export const TodoListItem = ({todo, index, handleDelete, handleToggle}) => {
    return (
        <li 
            key={todo.id} 
            className="list-group-item"
        >
            <p 
                className={`${todo.done && "complete"}`}
                onClick={ () =>handleToggle(todo.id) }    
            >
                {`${index}. ${todo.desc}` }
            </p>
            <button 
                type="button"
                className="btn btn-danger"
                onClick={ () => handleDelete(todo.id)}
            >
                Delete
            </button>

        </li>
    )
}

TodoListItem.propTypes = {
    todo: propTypes.object.isRequired,
    index: propTypes.number.isRequired,
    handleDelete: propTypes.func.isRequired,
    handleToggle: propTypes.func.isRequired
}
