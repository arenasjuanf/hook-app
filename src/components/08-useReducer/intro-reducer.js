const initialState = [{
    id:1,
    todo: "buy bread",
    done: false
}];

const todoReducer = (state = initialState, action) => {

    if(action?.type === "add"){

        return [...state, action.payload];
    }
    
    return state;
}

let todos = todoReducer();

const newTodo = {
    id:2,
    todo: "tidy desk",
    done: false
};

const addAction = {
    type: "add",
    payload: newTodo
};


todos = todoReducer(todos, addAction);


console.log(todos);

