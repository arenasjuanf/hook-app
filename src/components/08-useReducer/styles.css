body{
    height: 100%;
    padding: 50px;
}

p{
    cursor: pointer;
    margin: 0;
}

.list-group-item{
    align-items: center;
    border-bottom: 1px solid #e1dada !important;
    display: flex;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}

.add-title{
    border-radius: 5px;
    background-color: #08278094;
    color: white;
    padding: 5px;
}

.list-container{
    max-height: 500px;
    overflow-y: auto;
}