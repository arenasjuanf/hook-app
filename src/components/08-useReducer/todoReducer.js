export const todoReducer = (state = [], action) => {
    switch (action.type){
        case "add":
            return [...state, action.payload];
        case "delete":
            return state.filter( (t) => t.id !== action.payload );
        case "toggle":
            const pos = state.findIndex(t => t.id === action.payload);
            if( pos!== -1) state[pos].done = !state[pos].done;
            return [...state];
        default:
            return state;
    }

};
