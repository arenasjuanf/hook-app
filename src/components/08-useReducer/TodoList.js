import React from 'react';
import propTypes from 'prop-types';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, i) => (
                    <TodoListItem
                        key={todo.id}
                        todo={todo}
                        index={i+1}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                    
                ))
            }
        </ul>
    )
}
TodoList.propTypes = {
    todos: propTypes.array.isRequired,
    handleDelete: propTypes.func.isRequired,
    handleToggle: propTypes.func.isRequired
}