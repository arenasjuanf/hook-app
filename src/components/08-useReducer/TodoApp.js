import React, { useReducer, useEffect } from 'react'
import './styles.css';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem("todos")) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
       localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos])


    const handleDelete = (id) => {

        const action = {
            type: "delete",
            payload: id
        };

        dispatch(action);
    }

    const handleToggle = (id) => {

        const action = {
            type: "toggle",
            payload: id
        };

        dispatch(action);

    }

    const handleAddTodo = ( newTodo ) => {

        dispatch( {
            type: "add",
            payload: newTodo
        });
    }

    return (
        <div>   
            <h1>Todo App <small> ( {todos.length} ) </small></h1>
            <hr />
            
            <div className="row">
                <div className="col-7 list-container">
                    <TodoList 
                        todos={todos}
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />                    
                </div>

                <div className="col-5">
                    <TodoAdd 
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>
        </div>
    )
}
