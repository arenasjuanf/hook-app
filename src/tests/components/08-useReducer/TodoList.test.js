import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from "../../fixtures/demoTodos";

describe("<TodoList /> tests", () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(<TodoList
        todos={ demoTodos }
        handleDelete  = { handleDelete }
        handleToggle = { handleToggle }
    />);

    test('it should show succesfully', () => {

        expect(wrapper).toMatchSnapshot();

    })

    test('it should have two <TodoListItem/>', () => {
        expect(wrapper.find('TodoListItem').length).toBe(2);
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
    })
    
})