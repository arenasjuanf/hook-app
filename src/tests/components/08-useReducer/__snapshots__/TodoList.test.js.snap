// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> tests it should show succesfully 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="1"
    todo={
      Object {
        "desc": "learn react",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={2}
    key="2"
    todo={
      Object {
        "desc": "learn mongodb",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
